aiida/tools/importexport/archive/common.py

```diff
@@ -61,3 +78,384 @@ def detect_archive_type(in_path: str) -> str:
+    """Read a text based file from inside a tar file and return its content.
+
+    :param filepath: the path to the tar file
+    :param path: the relative path within the tar file
+
+    """
+    try:
+        with tarfile.open(filepath, 'r:*', format=tarfile.PAX_FORMAT) as handle:
+            result = handle.extractfile(path)
+            if result is None:
+                raise CorruptArchive(f'required file `{path}` is not included')
+            output = result.read()
+            if isinstance(output, bytes):
+                return output.decode('utf8')
+    except tarfile.ReadError:
+        raise ValueError('The input file format is not valid (not a tar file)')
+    except (KeyError, AttributeError):
+        raise CorruptArchive(f'required file `{path}` is not included')
+
+
+def _get_filter(only_prefix: Iterable[str], ignore_prefix: Iterable[str]) -> Callable[[str], bool]:
+    """Create filter for members to extract.
+
+    :param only_prefix: Extract only internal paths starting with these prefixes
+    :param ignore_prefix: Ignore internal paths starting with these prefixes
+
+    """
+    if only_prefix:
+
+        def _filter(name):
+            return any(name.startswith(prefix) for prefix in only_prefix
+                       ) and all(not name.startswith(prefix) for prefix in ignore_prefix)
+    else:
+
+        def _filter(name):
+            return all(not name.startswith(prefix) for prefix in ignore_prefix)
+
+    return _filter
+
+
+def safe_extract_zip(
+    in_path: Union[str, Path],
+    out_path: Union[str, Path],
+    *,
+    only_prefix: Iterable[str] = (),
+    ignore_prefix: Iterable[str] = ('..', '/'),
+    callback: Callable[[str, Any], None] = null_callback,
+    callback_description: str = 'Extracting zip files'
+):
+    """Safely extract a zip file
+
+    :param in_path: Path to extract from
+    :param out_path: Path to extract to
+    :param only_prefix: Extract only internal paths starting with these prefixes
+    :param ignore_prefix: Ignore internal paths starting with these prefixes
+    :param callback: a callback to report on the process, ``callback(action, value)``,
+        with the following callback signatures:
+
+        - ``callback('init', {'total': <int>, 'description': <str>})``,
+            to signal the start of a process, its total iterations and description
+        - ``callback('update', <int>)``,
+            to signal an update to the process and the number of iterations to progress
+
+    :param callback_description: the description to return in the callback
+
+    :raises `~aiida.tools.importexport.common.exceptions.CorruptArchive`: if the file cannot be read
+
+    """
+    _filter = _get_filter(only_prefix, ignore_prefix)
+    try:
+        with zipfile.ZipFile(in_path, 'r', allowZip64=True) as handle:
+            callback('init', {'total': 1, 'description': 'Computing zip objects to extract'})
+            members = [name for name in handle.namelist() if _filter(name)]
```

Comment:
Just to understand: Is it necessary/more performant to build this list first?
Why not just iterate over`handle.namelist()` and `continue` if the name doesn't fit
This way it's easier for users to follow the progress.

Reply:
It wouldn't make it easier to follow progress when you are providing a prefix filter (as is done in the archive reader / import code), whereby you may only be extracting a subfolder within the zip folder. Then the progress bar would be very misleading, because it might list 100,000s of objects to extract but actually only a small subset of them are really extracted.